File: desktop-app/src-tauri/src/quota/cloud_code.rs
use std::sync::{Mutex, OnceLock};
use std::time::{Duration, Instant};

use reqwest::Client;
use serde_json::{json, Value};

use crate::{CreditInfo, FullStatus, QuotaData};

use super::credentials::{find_agy_binary, load_credential, read_oauth_clients_from_binary};
use super::{is_usable_status, ProviderError, ProviderErrorKind};

const PROVIDER: &str = "Cloud Code";
const CLOUD_CODE_BASE: &str = "https://cloudcode-pa.googleapis.com";
const TOKEN_URL: &str = "https://oauth2.googleapis.com/token";
const HTTP_TIMEOUT: Duration = Duration::from_secs(12);
const PLAN_TIER_CACHE_TTL: Duration = Duration::from_secs(10 * 60);

static PLAN_TIER_CACHE: OnceLock<Mutex<Option<(String, Instant)>>> = OnceLock::new();

#[derive(Default, Clone)]
struct PoolValue {
    percent: u32,
    reset: String,
    present: bool,
}

#[derive(Default)]
struct ProviderPool {
    generic: PoolValue,
    five_hour: PoolValue,
    weekly: PoolValue,
}

fn plan_tier_cache() -> &'static Mutex<Option<(String, Instant)>> {
    PLAN_TIER_CACHE.get_or_init(|| Mutex::new(None))
}

fn extract_plan_tier(load: &Value) -> Option<String> {
    load.pointer("/paidTier/name")
        .or_else(|| load.pointer("/currentTier/name"))
        .or_else(|| load.pointer("/paidTier/id"))
        .or_else(|| load.pointer("/currentTier/id"))
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

fn cache_plan_tier(plan_tier: &str) {
    if let Ok(mut cache) = plan_tier_cache().lock() {
        *cache = Some((plan_tier.to_string(), Instant::now() + PLAN_TIER_CACHE_TTL));
    }
}

fn cached_plan_tier() -> Option<String> {
    let cache = plan_tier_cache().lock().ok()?;
    let (value, expires_at) = cache.as_ref()?;
    (Instant::now() < *expires_at).then(|| value.clone())
}

pub fn normalize_cloud_snapshot(
    load_code_assist: &Value,
    user_quota: &Value,
    available_models: &Value,
) -> Result<FullStatus, ProviderError> {
    let plan_tier = extract_plan_tier(load_code_assist);

    let credits = sum_available_credits(load_code_assist).map(|balance| CreditInfo {
        balance,
        credit_type: "AI".to_string(),
    });

    let mut gemini = ProviderPool::default();
    let mut shared = ProviderPool::default();

    if let Some(buckets) = user_quota.get("buckets").and_then(Value::as_array) {
        for bucket in buckets {
            let model = bucket
                .get("modelId")
                .or_else(|| bucket.get("model_id"))
                .and_then(Value::as_str)
                .unwrap_or("")
                .to_lowercase();
            let token_type = bucket
                .get("tokenType")
                .or_else(|| bucket.get("token_type"))
                .and_then(Value::as_str)
                .unwrap_or("")
                .to_lowercase();
            let target = if model.contains("gemini") || token_type == "requests" {
                Some(&mut gemini)
            } else if is_shared_model(&model) {
                Some(&mut shared)
            } else {
                None
            };
            let Some(target) = target else { continue };
            let Some(percent) = remaining_percent(bucket) else { continue };
            let reset = reset_time(bucket).unwrap_or_default();
            merge_pool(&mut target.generic, percent, reset.clone());
            match explicit_window(bucket).as_str() {
                "5h" => merge_pool(&mut target.five_hour, percent, reset),
                "weekly" => merge_pool(&mut target.weekly, percent, reset),
                _ => {}
            }
        }
    }

    if let Some(models) = available_models.get("models").and_then(Value::as_object) {
        for (id, entry) in models {
            let identity = format!(
                "{} {} {}",
                id,
                entry.get("model").and_then(Value::as_str).unwrap_or(""),
                entry.get("displayName").and_then(Value::as_str).unwrap_or("")
            )
            .to_lowercase();
            let target = if identity.contains("gemini") {
                Some(&mut gemini)
            } else if is_shared_model(&identity) {
                Some(&mut shared)
            } else {
                None
            };
            let Some(target) = target else { continue };
            let quota_info = entry.get("quotaInfo").or_else(|| entry.get("quota_info")).unwrap_or(entry);
            let Some(percent) = remaining_percent(quota_info) else { continue };
            let reset = reset_time(quota_info).unwrap_or_default();
            merge_pool(&mut target.generic, percent, reset.clone());
            match explicit_window(quota_info).as_str() {
                "5h" => merge_pool(&mut target.five_hour, percent, reset),
                "weekly" => merge_pool(&mut target.weekly, percent, reset),
                _ => {}
            }
        }
    }

    let has_pool_data = gemini.generic.present
        || gemini.five_hour.present
        || gemini.weekly.present
        || shared.generic.present
        || shared.five_hour.present
        || shared.weekly.present;
    let quotas = if has_pool_data {
        vec![build_card("Gemini", gemini), build_card("Claude & OpenAI", shared)]
    } else {
        Vec::new()
    };

    let status = FullStatus {
        credits,
        quotas,
        plan_tier,
        recently_used_model: None,
    };
    if !is_usable_status(&status) {
        return Err(ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, "Cloud Code returned no usable quota, credits, or tier data"));
    }
    Ok(status)
}

fn sum_available_credits(load: &Value) -> Option<f64> {
    let credits = load.pointer("/paidTier/availableCredits")?.as_array()?;
    let mut total = 0.0;
    let mut found = false;
    for credit in credits {
        let value = credit.get("creditAmount").or_else(|| credit.get("credit_amount"));
        let Some(amount) = value.and_then(number_value) else { continue };
        if amount.is_finite() && amount >= 0.0 {
            total += amount;
            found = true;
        }
    }
    found.then_some(total)
}

fn is_shared_model(value: &str) -> bool {
    value.contains("claude") || value.contains("gpt") || value.contains("openai")
}

fn remaining_percent(value: &Value) -> Option<u32> {
    for key in ["remainingFraction", "remaining_fraction"] {
        if let Some(number) = value.get(key).and_then(number_value) {
            return Some((number.clamp(0.0, 1.0) * 100.0).round() as u32);
        }
    }
    for key in ["remainingPercent", "remaining_percent", "percentRemaining", "percent_remaining"] {
        if let Some(number) = value.get(key).and_then(number_value) {
            return Some(number.clamp(0.0, 100.0).round() as u32);
        }
    }
    None
}

fn number_value(value: &Value) -> Option<f64> {
    value.as_f64().or_else(|| value.as_str()?.replace(',', "").parse::<f64>().ok())
}

fn reset_time(value: &Value) -> Option<String> {
    value
        .get("resetTime")
        .or_else(|| value.get("reset_time"))
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string)
}

fn explicit_window(value: &Value) -> String {
    let text = value
        .get("window")
        .or_else(|| value.get("windowName"))
        .or_else(|| value.get("window_name"))
        .and_then(Value::as_str)
        .unwrap_or("")
        .trim()
        .to_lowercase();
    let compact: String = text.chars().filter(|c| !c.is_whitespace() && *c != '-' && *c != '_').collect();
    if compact == "5h" || compact.contains("5hour") || compact.contains("fivehour") {
        "5h".to_string()
    } else if compact.contains("week") || compact == "7d" {
        "weekly".to_string()
    } else {
        String::new()
    }
}

fn merge_pool(slot: &mut PoolValue, percent: u32, reset: String) {
    if !slot.present || percent < slot.percent || (percent == slot.percent && !reset.is_empty() && (slot.reset.is_empty() || reset < slot.reset)) {
        slot.percent = percent;
        slot.reset = reset;
        slot.present = true;
    }
}

fn build_card(model: &str, pool: ProviderPool) -> QuotaData {
    let generic = if pool.generic.present {
        pool.generic.clone()
    } else if pool.five_hour.present {
        pool.five_hour.clone()
    } else {
        pool.weekly.clone()
    };
    QuotaData {
        model: model.to_string(),
        percent: generic.percent,
        refresh_time: generic.reset,
        five_hour_percent: pool.five_hour.percent,
        five_hour_reset: pool.five_hour.reset,
        five_hour_disabled: !pool.five_hour.present,
        weekly_percent: pool.weekly.percent,
        weekly_reset: pool.weekly.reset,
        weekly_disabled: !pool.weekly.present,
    }
}

async fn build_authenticated_client() -> Result<(Client, String), ProviderError> {
    let credential = load_credential().await?;
    let agy = find_agy_binary()
        .ok_or_else(|| ProviderError::new(PROVIDER, ProviderErrorKind::Unavailable, "agy executable was not found for OAuth client discovery"))?;
    let candidates = read_oauth_clients_from_binary(&agy)?;
    let client = Client::builder()
        .timeout(HTTP_TIMEOUT)
        .build()
        .map_err(|_| ProviderError::new(PROVIDER, ProviderErrorKind::Transient, "could not initialize HTTPS client"))?;
    let access_token = refresh_access_token(&client, &credential.refresh_token, &candidates).await?;
    Ok((client, access_token))
}

pub async fn fetch_plan_tier() -> Result<Option<String>, ProviderError> {
    if let Some(plan_tier) = cached_plan_tier() {
        return Ok(Some(plan_tier));
    }

    let (client, access_token) = build_authenticated_client().await?;
    let load = post_cloud_json(
        &client,
        "loadCodeAssist",
        &access_token,
        json!({
            "metadata": cloud_metadata(),
            "mode": "FULL_ELIGIBILITY_CHECK"
        }),
        false,
    )
    .await?;

    let plan_tier = extract_plan_tier(&load);
    if let Some(value) = plan_tier.as_deref() {
        cache_plan_tier(value);
    }
    Ok(plan_tier)
}

pub async fn fetch() -> Result<FullStatus, ProviderError> {
    let (client, access_token) = build_authenticated_client().await?;

    let load = post_cloud_json(
        &client,
        "loadCodeAssist",
        &access_token,
        json!({
            "metadata": cloud_metadata(),
            "mode": "FULL_ELIGIBILITY_CHECK"
        }),
        false,
    )
    .await
    .unwrap_or(Value::Null);

    if let Some(plan_tier) = extract_plan_tier(&load) {
        cache_plan_tier(&plan_tier);
    }

    let user_quota = post_cloud_json(&client, "retrieveUserQuota", &access_token, json!({}), false)
        .await
        .unwrap_or(Value::Null);

    let project = extract_project_id(&load);
    let models_body = project.map(|project| json!({"project": project})).unwrap_or_else(|| json!({}));
    let models = post_cloud_json(&client, "fetchAvailableModels", &access_token, models_body, true)
        .await
        .unwrap_or(Value::Null);

    normalize_cloud_snapshot(&load, &user_quota, &models)
}

async fn refresh_access_token(
    client: &Client,
    refresh_token: &str,
    candidates: &[(String, String)],
) -> Result<String, ProviderError> {
    for (client_id, client_secret) in candidates {
        let response = client
            .post(TOKEN_URL)
            .form(&[
                ("client_id", client_id.as_str()),
                ("client_secret", client_secret.as_str()),
                ("refresh_token", refresh_token),
                ("grant_type", "refresh_token"),
            ])
            .send()
            .await;
        let Ok(response) = response else { continue };
        if !response.status().is_success() {
            continue;
        }
        let Ok(value) = response.json::<Value>().await else { continue };
        if let Some(token) = value.get("access_token").and_then(Value::as_str).filter(|s| !s.is_empty()) {
            return Ok(token.to_string());
        }
    }
    Err(ProviderError::new(PROVIDER, ProviderErrorKind::Auth, "could not refresh agy OAuth access token"))
}

async fn post_cloud_json(
    client: &Client,
    endpoint: &str,
    access_token: &str,
    body: Value,
    antigravity_headers: bool,
) -> Result<Value, ProviderError> {
    let url = format!("{CLOUD_CODE_BASE}/v1internal:{endpoint}");
    let mut request = client
        .post(url)
        .bearer_auth(access_token)
        .header("Content-Type", "application/json")
        .json(&body);
    if antigravity_headers {
        request = request
            .header("User-Agent", "antigravity-quota-quickcheck/1.2.3")
            .header("X-Goog-Api-Client", "google-cloud-sdk vscode_cloudshelleditor/0.1")
            .header("Client-Metadata", serde_json::to_string(&cloud_metadata()).unwrap_or_default());
    }
    let response = request
        .send()
        .await
        .map_err(|_| ProviderError::new(PROVIDER, ProviderErrorKind::Transient, format!("{endpoint} request failed")))?;
    if response.status().as_u16() == 401 || response.status().as_u16() == 403 {
        return Err(ProviderError::new(PROVIDER, ProviderErrorKind::Auth, format!("{endpoint} rejected agy credentials")));
    }
    if !response.status().is_success() {
        return Err(ProviderError::new(PROVIDER, ProviderErrorKind::Transient, format!("{endpoint} returned HTTP {}", response.status().as_u16())));
    }
    response
        .json::<Value>()
        .await
        .map_err(|_| ProviderError::new(PROVIDER, ProviderErrorKind::InvalidData, format!("{endpoint} returned invalid JSON")))
}

fn cloud_metadata() -> Value {
    let platform = if cfg!(target_os = "windows") {
        "WINDOWS"
    } else if cfg!(target_os = "macos") {
        "DARWIN"
    } else {
        "LINUX"
    };
    json!({
        "ideName": "antigravity",
        "ideType": "ANTIGRAVITY",
        "platform": platform,
        "pluginType": "GEMINI",
        "updateChannel": "stable"
    })
}

fn extract_project_id(load: &Value) -> Option<String> {
    let value = load.get("cloudaicompanionProject")?;
    if let Some(text) = value.as_str().filter(|s| !s.is_empty()) {
        return Some(text.to_string());
    }
    for key in ["projectId", "project_id", "id", "name"] {
        if let Some(text) = value.get(key).and_then(Value::as_str).filter(|s| !s.is_empty()) {
            return Some(text.to_string());
        }
    }
    None
}

File: src/telemetry/providers/cloudCode.ts
import { CreditInfo, FullStatus, QuotaData } from '../types';
import {
  AgyCredential,
  OauthClient,
  loadAgyCredential,
  readOauthClientsFromAgy,
} from './credentials';
import { HttpStatusError, requestJson } from './http';
import { ProviderError, isUsableStatus } from './types';

const PROVIDER = 'Cloud Code';
const CLOUD_CODE_BASE = 'https://cloudcode-pa.googleapis.com';
const TOKEN_URL = 'https://oauth2.googleapis.com/token';
const MISSING_COOLDOWN_MS = 60_000;
const TRANSIENT_COOLDOWN_MS = 20_000;

interface PoolValue {
  percent: number;
  reset: string;
  present: boolean;
}

interface ProviderPool {
  generic: PoolValue;
  fiveHour: PoolValue;
  weekly: PoolValue;
}

interface RefreshResult {
  accessToken: string;
  refreshToken: string | null;
}

let missingUntil = 0;
let transientUntil = 0;
let workingClient: OauthClient | null = null;
let rotatedRefreshToken: string | null = null;

function emptyPoolValue(): PoolValue {
  return { percent: 0, reset: '', present: false };
}

function emptyProviderPool(): ProviderPool {
  return {
    generic: emptyPoolValue(),
    fiveHour: emptyPoolValue(),
    weekly: emptyPoolValue(),
  };
}

function numberValue(value: unknown): number | null {
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (typeof value === 'string' && value.trim()) {
    const parsed = Number(value.replace(/,/g, ''));
    return Number.isFinite(parsed) ? parsed : null;
  }
  return null;
}

function remainingPercent(value: any): number | null {
  for (const key of ['remainingFraction', 'remaining_fraction']) {
    const number = numberValue(value?.[key]);
    if (number !== null) {
      return Math.round(Math.max(0, Math.min(1, number)) * 100);
    }
  }

  for (const key of [
    'remainingPercent',
    'remaining_percent',
    'percentRemaining',
    'percent_remaining',
  ]) {
    const number = numberValue(value?.[key]);
    if (number !== null) {
      return Math.round(Math.max(0, Math.min(100, number)));
    }
  }

  return null;
}

function resetTime(value: any): string {
  const reset = value?.resetTime ?? value?.reset_time;
  return typeof reset === 'string' ? reset : '';
}

function explicitWindow(value: any): '5h' | 'weekly' | null {
  const raw = value?.window ?? value?.windowName ?? value?.window_name;
  if (typeof raw !== 'string') return null;
  const compact = raw.trim().toLowerCase().replace(/[\s_-]/g, '');
  if (compact === '5h' || compact.includes('5hour') || compact.includes('fivehour')) {
    return '5h';
  }
  if (compact.includes('week') || compact === '7d') {
    return 'weekly';
  }
  return null;
}

function mergePool(slot: PoolValue, percent: number, reset: string): void {
  if (
    !slot.present ||
    percent < slot.percent ||
    (percent === slot.percent && reset && (!slot.reset || reset < slot.reset))
  ) {
    slot.percent = percent;
    slot.reset = reset;
    slot.present = true;
  }
}

function isSharedModel(value: string): boolean {
  return value.includes('claude') || value.includes('gpt') || value.includes('openai');
}

function buildCard(model: string, pool: ProviderPool): QuotaData {
  const generic = pool.generic.present
    ? pool.generic
    : pool.fiveHour.present
      ? pool.fiveHour
      : pool.weekly;

  return {
    model,
    percent: generic.percent,
    refreshTime: generic.reset,
    fiveHourPercent: pool.fiveHour.percent,
    fiveHourReset: pool.fiveHour.reset,
    fiveHourDisabled: !pool.fiveHour.present,
    weeklyPercent: pool.weekly.percent,
    weeklyReset: pool.weekly.reset,
    weeklyDisabled: !pool.weekly.present,
  };
}

function sumAvailableCredits(load: any): CreditInfo | null {
  const credits = load?.paidTier?.availableCredits;
  if (!Array.isArray(credits)) return null;

  let total = 0;
  let found = false;
  for (const credit of credits) {
    const amount = numberValue(credit?.creditAmount ?? credit?.credit_amount);
    if (amount !== null && amount >= 0) {
      total += amount;
      found = true;
    }
  }

  return found ? { balance: total, creditType: 'AI' } : null;
}

export function parseCloudCodeStatus(
  loadCodeAssist: any,
  userQuota: any,
  availableModels: any,
): FullStatus {
  const tierCandidate =
    loadCodeAssist?.paidTier?.name ??
    loadCodeAssist?.currentTier?.name ??
    loadCodeAssist?.paidTier?.id ??
    loadCodeAssist?.currentTier?.id;
  const planTier = typeof tierCandidate === 'string' && tierCandidate
    ? tierCandidate
    : null;
  const credits = sumAvailableCredits(loadCodeAssist);

  const gemini = emptyProviderPool();
  const shared = emptyProviderPool();

  if (Array.isArray(userQuota?.buckets)) {
    for (const bucket of userQuota.buckets) {
      const model = String(bucket?.modelId ?? bucket?.model_id ?? '').toLowerCase();
      const tokenType = String(bucket?.tokenType ?? bucket?.token_type ?? '').toLowerCase();
      const target = model.includes('gemini') || tokenType === 'requests'
        ? gemini
        : isSharedModel(model)
          ? shared
          : null;
      if (!target) continue;

      const percent = remainingPercent(bucket);
      if (percent === null) continue;
      const reset = resetTime(bucket);
      mergePool(target.generic, percent, reset);
      const window = explicitWindow(bucket);
      if (window === '5h') mergePool(target.fiveHour, percent, reset);
      if (window === 'weekly') mergePool(target.weekly, percent, reset);
    }
  }

  if (availableModels?.models && typeof availableModels.models === 'object') {
    for (const [id, entryValue] of Object.entries(availableModels.models)) {
      const entry: any = entryValue;
      const identity = [id, entry?.model ?? '', entry?.displayName ?? '']
        .join(' ')
        .toLowerCase();
      const target = identity.includes('gemini')
        ? gemini
        : isSharedModel(identity)
          ? shared
          : null;
      if (!target) continue;

      const quotaInfo = entry?.quotaInfo ?? entry?.quota_info ?? entry;
      const percent = remainingPercent(quotaInfo);
      if (percent === null) continue;
      const reset = resetTime(quotaInfo);
      mergePool(target.generic, percent, reset);
      const window = explicitWindow(quotaInfo);
      if (window === '5h') mergePool(target.fiveHour, percent, reset);
      if (window === 'weekly') mergePool(target.weekly, percent, reset);
    }
  }

  const hasPoolData = [
    gemini.generic,
    gemini.fiveHour,
    gemini.weekly,
    shared.generic,
    shared.fiveHour,
    shared.weekly,
  ].some((value) => value.present);

  const status: FullStatus = {
    credits,
    quotas: hasPoolData
      ? [buildCard('Gemini', gemini), buildCard('Claude & OpenAI', shared)]
      : [],
    recentlyUsedModel: null,
    planTier,
  };

  if (!isUsableStatus(status)) {
    throw new ProviderError(
      PROVIDER,
      'invalid-data',
      'Cloud Code returned no usable quota, credits, or tier data',
    );
  }

  return status;
}

function cloudMetadata(): Record<string, string> {
  const platform = process.platform === 'win32'
    ? 'WINDOWS'
    : process.platform === 'darwin'
      ? 'DARWIN'
      : 'LINUX';
  return {
    ideName: 'antigravity',
    ideType: 'ANTIGRAVITY',
    platform,
    pluginType: 'GEMINI',
    updateChannel: 'stable',
  };
}

function classifyTransportError(endpoint: string, error: unknown): ProviderError {
  if (error instanceof ProviderError) return error;
  if (error instanceof HttpStatusError) {
    if (error.statusCode === 401 || error.statusCode === 403) {
      return new ProviderError(PROVIDER, 'auth', `${endpoint} rejected agy credentials`);
    }
    return new ProviderError(
      PROVIDER,
      'transient',
      `${endpoint} returned HTTP ${error.statusCode}`,
    );
  }
  const message = error instanceof Error ? error.message.toLowerCase() : '';
  const kind = message.includes('valid json') ? 'invalid-data' : 'transient';
  return new ProviderError(PROVIDER, kind, `${endpoint} request failed`);
}

async function refreshWithClients(
  refreshToken: string,
  clients: OauthClient[],
): Promise<RefreshResult> {
  let sawTransient = false;

  for (const candidate of clients) {
    const form = new URLSearchParams({
      client_id: candidate.clientId,
      client_secret: candidate.clientSecret,
      refresh_token: refreshToken,
      grant_type: 'refresh_token',
    }).toString();

    try {
      const value = await requestJson(TOKEN_URL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: form,
      });
      const accessToken = typeof value?.access_token === 'string'
        ? value.access_token
        : '';
      if (!accessToken) continue;

      workingClient = candidate;
      return {
        accessToken,
        refreshToken: typeof value?.refresh_token === 'string' && value.refresh_token
          ? value.refresh_token
          : null,
      };
    } catch (error) {
      if (error instanceof HttpStatusError) {
        if (error.statusCode >= 500 || error.statusCode === 429) {
          sawTransient = true;
        }
        continue;
      }
      sawTransient = true;
    }
  }

  throw new ProviderError(
    PROVIDER,
    sawTransient ? 'transient' : 'auth',
    sawTransient
      ? 'OAuth refresh request failed'
      : 'could not refresh agy OAuth access token',
  );
}

async function refreshAccessToken(
  credential: AgyCredential,
): Promise<string> {
  let candidates: OauthClient[];
  if (workingClient) {
    candidates = [workingClient];
  } else {
    candidates = await readOauthClientsFromAgy();
  }

  const preferredRefreshToken = rotatedRefreshToken ?? credential.refreshToken;
  try {
    const result = await refreshWithClients(preferredRefreshToken, candidates);
    if (result.refreshToken) rotatedRefreshToken = result.refreshToken;
    return result.accessToken;
  } catch (error) {
    if (workingClient) {
      workingClient = null;
      candidates = await readOauthClientsFromAgy();
      const retryToken = preferredRefreshToken === credential.refreshToken
        ? credential.refreshToken
        : credential.refreshToken;
      const result = await refreshWithClients(retryToken, candidates);
      if (result.refreshToken) rotatedRefreshToken = result.refreshToken;
      return result.accessToken;
    }
    if (rotatedRefreshToken && preferredRefreshToken !== credential.refreshToken) {
      rotatedRefreshToken = null;
      const result = await refreshWithClients(credential.refreshToken, candidates);
      if (result.refreshToken) rotatedRefreshToken = result.refreshToken;
      return result.accessToken;
    }
    throw error;
  }
}

async function postCloudJson(
  endpoint: string,
  accessToken: string,
  body: unknown,
): Promise<any> {
  const metadata = cloudMetadata();
  try {
    return await requestJson(`${CLOUD_CODE_BASE}/v1internal:${endpoint}`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
        'User-Agent': 'antigravity-quota-quickcheck-vscode',
        'X-Goog-Api-Client': 'google-cloud-sdk vscode_cloudshelleditor/0.1',
        'Client-Metadata': JSON.stringify(metadata),
      },
      body,
    });
  } catch (error) {
    throw classifyTransportError(endpoint, error);
  }
}

function extractProjectId(load: any): string | null {
  const value = load?.cloudaicompanionProject;
  if (typeof value === 'string' && value) return value;
  if (!value || typeof value !== 'object') return null;
  for (const key of ['projectId', 'project_id', 'id', 'name']) {
    const candidate = value[key];
    if (typeof candidate === 'string' && candidate) return candidate;
  }
  return null;
}

async function optionalCloudCall(
  endpoint: string,
  accessToken: string,
  body: unknown,
): Promise<any> {
  try {
    return await postCloudJson(endpoint, accessToken, body);
  } catch (error) {
    if (error instanceof ProviderError && error.kind === 'auth') throw error;
    return null;
  }
}

export async function fetchCloudCode(force: boolean): Promise<FullStatus> {
  const now = Date.now();
  if (now < missingUntil) {
    throw new ProviderError(PROVIDER, 'unavailable', 'agy credentials are temporarily unavailable');
  }
  if (!force && now < transientUntil) {
    throw new ProviderError(PROVIDER, 'transient', 'Cloud Code is cooling down after a transient failure');
  }

  try {
    let credential: AgyCredential;
    try {
      credential = await loadAgyCredential();
    } catch (error) {
      if (error instanceof ProviderError && (error.kind === 'unavailable' || error.kind === 'unsupported')) {
        missingUntil = Date.now() + MISSING_COOLDOWN_MS;
      }
      throw error;
    }

    let accessToken: string;
    try {
      accessToken = await refreshAccessToken(credential);
    } catch (error) {
      if (error instanceof ProviderError && (error.kind === 'unavailable' || error.kind === 'unsupported')) {
        missingUntil = Date.now() + MISSING_COOLDOWN_MS;
      }
      throw error;
    }

    const metadata = cloudMetadata();
    const [load, quota] = await Promise.all([
      optionalCloudCall('loadCodeAssist', accessToken, {
        metadata,
        mode: 'FULL_ELIGIBILITY_CHECK',
      }),
      optionalCloudCall('retrieveUserQuota', accessToken, {}),
    ]);

    const project = extractProjectId(load);
    const models = await optionalCloudCall(
      'fetchAvailableModels',
      accessToken,
      project ? { project } : {},
    );

    return parseCloudCodeStatus(load, quota, models);
  } catch (error) {
    const providerError = error instanceof ProviderError
      ? error
      : new ProviderError(PROVIDER, 'transient', 'Cloud Code provider failed');
    if (providerError.kind === 'transient') {
      transientUntil = Date.now() + TRANSIENT_COOLDOWN_MS;
    }
    throw providerError;
  }
}
